pkg/rules/service.go

```diff
@@ -0,0 +1,347 @@
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	if s.ruleFiles == nil {
+		keys := make([]string, 0)
+		for k := range cm.Data {
+			keys = append(keys, k)
+		}
+		if len(keys) > 0 {
+			s.ruleFiles = keys
+		} else {
+			return fmt.Errorf("rule with ID %s not found", id)
+		}
+	}
+
+	for _, fileName := range s.ruleFiles {
+		rules := s.extractRules(cm.Data, fileName)
+
+		if rules == nil {
+			continue
+		}
+
+		var checkRuleIncluded *RuleData
+
+		for i, r := range rules {
+			if r.Id == id {
+				rules[i] = &newRule
+				checkRuleIncluded = r
+			}
+		}
+
+		if checkRuleIncluded == nil {
+			continue
+		}
+
+		rawRules, err := json.Marshal(rules)
+
+		if err != nil {
+			return err
+		}
+
+		cm.Data[fileName] = string(rawRules)
+
+		if _, err = s.k8s.ConfigMaps(s.cmNamespace).Update(ctx, cm, metaV1.UpdateOptions{}); err != nil {
+
+			return err
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("rule with ID %s not found", id)
+}
+
+func (s *Service) CreateRule(ctx context.Context, id string, newRule RuleData, userFile string) error {
+	ctx, span := s.tracer.Start(ctx, "rules.Service.CreateRule")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	if s.ruleFiles == nil {
+		keys := make([]string, 0)
+		for k := range cm.Data {
+			keys = append(keys, k)
+		}
+		if !contains(keys, "admin-ui-rules.json") {
+			keys = append(keys, "admin-ui-rules.json")
+		}
+		s.ruleFiles = keys
+	}
+
+	if userFile == "" {
+		userFile = "admin-ui-rules.json"
+	}
+
+	for _, fileName := range s.ruleFiles {
+
+		rules := s.extractRules(cm.Data, fileName)
+
+		if rules == nil {
+			continue
+		}
+
+		for _, r := range rules {
+			if r.Id == id {
+				return fmt.Errorf("rule with ID %s already exists", id)
+			}
+		}
+	}
+
+	var rules []*RuleData
+
+	_, ok := cm.Data[userFile]
+	if ok {
+		rules = s.extractRules(cm.Data, userFile)
+		rules = append(rules, &newRule)
+	} else {
+		rules = []*RuleData{&newRule}
+	}
+
+	rawRules, err := json.Marshal(rules)
+
+	if err != nil {
+		return err
+	}
+
+	cm.Data[userFile] = string(rawRules)
+
+	if _, err = s.k8s.ConfigMaps(s.cmNamespace).Update(ctx, cm, metaV1.UpdateOptions{}); err != nil {
+
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) DeleteRule(ctx context.Context, id string) error {
+	ctx, span := s.tracer.Start(ctx, "rules.Service.DeleteRule")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	found := false
+
+	if s.ruleFiles == nil {
+		keys := make([]string, 0)
+		for k := range cm.Data {
+			keys = append(keys, k)
+		}
+		if len(keys) > 0 {
+			s.ruleFiles = keys
+		} else {
+			return fmt.Errorf("rule with ID %s not found", id)
+		}
+	}
```

Comment:
I don't think this is needed, also AFAICT it is not correct as it assumes that only this process will be making changes to the ruleFiles.

Reply:
This code is now removed.